File: src/utils/security/encryption.ts
/**
 * AES-256-GCM Encryption Utilities
 *
 * Provides client-side encryption/decryption using Web Crypto API.
 * Used for Tier 2 analytics to encrypt data before sending to backend.
 *
 * Security Features:
 * - AES-256-GCM (authenticated encryption)
 * - Random IV for each encryption
 * - Authentication tags verified on decryption
 * - Key derivation using PBKDF2 (100k iterations)
 *
 * @module encryption
 */

import logger from "@/utils/core/common/logger";

/**
 * Encryption result containing all necessary components
 */
export interface EncryptedData {
  /** Base64-encoded encrypted data */
  ciphertext: string;
  /** Base64-encoded initialization vector */
  iv: string;
  /** Base64-encoded authentication tag */
  authTag: string;
  /** Algorithm used (always 'AES-256-GCM') */
  algorithm: string;
}

/**
 * Decryption input (mirrors EncryptedData)
 */
export interface EncryptedPayload {
  ciphertext: string;
  iv: string;
  authTag: string;
  algorithm: string;
}

/**
 * Encryption configuration
 */
const ENCRYPTION_CONFIG = {
  algorithm: "AES-GCM" as const,
  keyLength: 256, // bits
  ivLength: 12, // bytes (96 bits recommended for GCM)
  tagLength: 128, // bits
} as const;

/**
 * Convert ArrayBuffer to Base64 string
 */
function arrayBufferToBase64(buffer: ArrayBuffer): string {
  const bytes = new Uint8Array(buffer);
  let binary = "";
  for (let i = 0; i < bytes.length; i++) {
    binary += String.fromCharCode(bytes[i]);
  }
  return btoa(binary);
}

/**
 * Convert Base64 string to ArrayBuffer
 */
function base64ToArrayBuffer(base64: string): ArrayBuffer {
  const binary = atob(base64);
  const bytes = new Uint8Array(binary.length);
  for (let i = 0; i < binary.length; i++) {
    bytes[i] = binary.charCodeAt(i);
  }
  return bytes.buffer;
}

/**
 * Generate a random initialization vector (IV)
 *
 * @returns 96-bit random IV as Uint8Array
 */
export function generateIV(): Uint8Array {
  return crypto.getRandomValues(new Uint8Array(ENCRYPTION_CONFIG.ivLength));
}

/**
 * Encrypt data using AES-256-GCM
 *
 * @param data - Data to encrypt (will be JSON stringified)
 * @param key - CryptoKey for encryption (256-bit AES-GCM key)
 * @returns Encrypted data with IV and auth tag
 *
 * @example
 * ```typescript
 * const key = await deriveKey(passphrase, salt);
 * const encrypted = await encryptData({ userId: 123, amount: 50.00 }, key);
 * // Send encrypted.ciphertext, encrypted.iv, encrypted.authTag to backend
 * ```
 */
export async function encryptData(data: unknown, key: CryptoKey): Promise<EncryptedData> {
  try {
    // Serialize data to JSON
    const plaintext = JSON.stringify(data);
    const encoder = new TextEncoder();
    const plaintextBytes = encoder.encode(plaintext);

    // Generate random IV
    const iv = generateIV();

    // Encrypt
    const ciphertextWithTag = await crypto.subtle.encrypt(
      {
        name: ENCRYPTION_CONFIG.algorithm,
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        iv: iv as any,
        tagLength: ENCRYPTION_CONFIG.tagLength,
      },
      key,
      plaintextBytes
    );

    // GCM auth tag is appended to ciphertext
    const ciphertext = ciphertextWithTag.slice(0, -16); // Last 16 bytes are auth tag
    const authTag = ciphertextWithTag.slice(-16);

    return {
      ciphertext: arrayBufferToBase64(ciphertext),
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      iv: arrayBufferToBase64(iv as any),
      authTag: arrayBufferToBase64(authTag),
      algorithm: "AES-256-GCM",
    };
  } catch (error) {
    logger.error("Encryption failed", { error });
    throw new Error("Failed to encrypt data");
  }
}

/**
 * Decrypt data using AES-256-GCM
 *
 * @param encrypted - Encrypted data with IV and auth tag
 * @param key - CryptoKey for decryption (same key used for encryption)
 * @returns Decrypted data (parsed from JSON)
 *
 * @throws Error if authentication tag verification fails
 *
 * @example
 * ```typescript
 * const key = await deriveKey(passphrase, salt);
 * const decrypted = await decryptData(encryptedPayload, key);
 * console.log(decrypted); // { userId: 123, amount: 50.00 }
 * ```
 */
export async function decryptData<T = unknown>(
  encrypted: EncryptedPayload,
  key: CryptoKey
): Promise<T> {
  try {
    // Verify algorithm
    if (encrypted.algorithm !== "AES-256-GCM") {
      throw new Error(`Unsupported algorithm: ${encrypted.algorithm}`);
    }

    // Convert from base64
    const ciphertext = base64ToArrayBuffer(encrypted.ciphertext);
    const iv = base64ToArrayBuffer(encrypted.iv);
    const authTag = base64ToArrayBuffer(encrypted.authTag);

    // Combine ciphertext and auth tag (GCM expects them concatenated)
    const ciphertextWithTag = new Uint8Array(ciphertext.byteLength + authTag.byteLength);
    ciphertextWithTag.set(new Uint8Array(ciphertext), 0);
    ciphertextWithTag.set(new Uint8Array(authTag), ciphertext.byteLength);

    // Decrypt
    const plaintextBytes = await crypto.subtle.decrypt(
      {
        name: ENCRYPTION_CONFIG.algorithm,
        iv: new Uint8Array(iv),
        tagLength: ENCRYPTION_CONFIG.tagLength,
      },
      key,
      ciphertextWithTag
    );

    // Decode and parse
    const decoder = new TextDecoder();
    const plaintext = decoder.decode(plaintextBytes);
    return JSON.parse(plaintext) as T;
  } catch (error) {
    logger.error("Decryption failed", { error });
    throw new Error("Failed to decrypt data (authentication failed or corrupted data)");
  }
}

/**
 * Derive encryption key from passphrase using PBKDF2
 *
 * @param passphrase - User passphrase (never persisted)
 * @param salt - Salt for key derivation (base64 encoded)
 * @param iterations - PBKDF2 iterations (default: 100,000)
 * @returns CryptoKey for AES-GCM encryption
 *
 * @example
 * ```typescript
 * const salt = generateSalt();
 * const key = await deriveKey("my-secret-passphrase", salt);
 * ```
 */
export async function deriveKey(
  passphrase: string,
  salt: string,
  iterations: number = 100_000
): Promise<CryptoKey> {
  try {
    const encoder = new TextEncoder();
    const passphraseBytes = encoder.encode(passphrase);
    const saltBytes = base64ToArrayBuffer(salt);

    // Import passphrase as key material
    const keyMaterial = await crypto.subtle.importKey("raw", passphraseBytes, "PBKDF2", false, [
      "deriveKey",
    ]);

    // Derive AES-GCM key
    const key = await crypto.subtle.deriveKey(
      {
        name: "PBKDF2",
        salt: saltBytes,
        iterations,
        hash: "SHA-256",
      },
      keyMaterial,
      {
        name: ENCRYPTION_CONFIG.algorithm,
        length: ENCRYPTION_CONFIG.keyLength,
      },
      false, // Not extractable (stays in memory)
      ["encrypt", "decrypt"]
    );

    return key;
  } catch (error) {
    logger.error("Key derivation failed", { error });
    throw new Error("Failed to derive encryption key");
  }
}

/**
 * Generate a random salt for PBKDF2
 *
 * @returns Base64-encoded salt (128 bits)
 */
export function generateSalt(): string {
  const salt = crypto.getRandomValues(new Uint8Array(16)); // 128 bits
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  return arrayBufferToBase64(salt as any);
}

/**
 * Export key for storage (wrapped with another key)
 *
 * NOT RECOMMENDED: Keys should stay in memory only.
 * This is provided for advanced use cases only.
 *
 * @param key - CryptoKey to export
 * @param wrappingKey - CryptoKey to wrap with
 * @returns Base64-encoded wrapped key
 */
export async function exportKey(key: CryptoKey, wrappingKey: CryptoKey): Promise<string> {
  try {
    const wrapped = await crypto.subtle.wrapKey("raw", key, wrappingKey, {
      name: ENCRYPTION_CONFIG.algorithm,
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
      iv: generateIV() as any,
    });
    return arrayBufferToBase64(wrapped);
  } catch (error) {
    logger.error("Key export failed", { error });
    throw new Error("Failed to export key");
  }
}

/**
 * Clear sensitive data from memory
 *
 * Note: JavaScript doesn't provide direct memory management,
 * but this helps signal to the GC that data can be collected.
 *
 * @param sensitiveData - Data to clear (object or string)
 */
export function clearSensitiveData(sensitiveData: Record<string, unknown> | string): void {
  if (typeof sensitiveData === "object") {
    // Overwrite all properties
    for (const key in sensitiveData) {
      if (Object.prototype.hasOwnProperty.call(sensitiveData, key)) {
        sensitiveData[key] = null;
      }
    }
  }

  // Force garbage collection hint (not guaranteed)
  if (typeof (globalThis as typeof globalThis & { gc?: () => void }).gc === "function") {
    (globalThis as typeof globalThis & { gc?: () => void }).gc?.();
  }
}

/**
 * Test encryption round-trip
 *
 * @param data - Test data to encrypt and decrypt
 * @param key - CryptoKey to use
 * @returns True if round-trip successful, false otherwise
 *
 * @example
 * ```typescript
 * const key = await deriveKey("test", generateSalt());
 * const success = await testEncryption({ test: "data" }, key);
 * console.log(success ? "Encryption working" : "Encryption failed");
 * ```
 */
export async function testEncryption(data: unknown, key: CryptoKey): Promise<boolean> {
  try {
    const encrypted = await encryptData(data, key);
    const decrypted = await decryptData(encrypted, key);

    // Deep equality check
    return JSON.stringify(data) === JSON.stringify(decrypted);
  } catch (error) {
    logger.error("Encryption test failed", { error });
    return false;
  }
}

File: api/analytics/utils/encryption.go
package utils

import (
	"crypto/aes"
	"crypto/cipher"
	"crypto/rand"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"runtime"
)

// DecryptPayload decrypts an AES-256-GCM encrypted payload
//
// The payload should contain:
// - Ciphertext (base64 encoded)
// - IV (base64 encoded initialization vector)
// - AuthTag (base64 encoded authentication tag)
// - Algorithm (should be "AES-256-GCM")
//
// Returns the decrypted data as a byte slice that can be unmarshaled into the desired type.
func DecryptPayload(encrypted EncryptedPayload, key []byte) ([]byte, error) {
	// Verify algorithm
	if encrypted.Algorithm != "AES-256-GCM" {
		return nil, fmt.Errorf("unsupported algorithm: %s", encrypted.Algorithm)
	}

	// Decode base64 strings
	ciphertext, err := base64.StdEncoding.DecodeString(encrypted.Ciphertext)
	if err != nil {
		return nil, fmt.Errorf("failed to decode ciphertext: %w", err)
	}

	iv, err := base64.StdEncoding.DecodeString(encrypted.IV)
	if err != nil {
		return nil, fmt.Errorf("failed to decode IV: %w", err)
	}

	authTag, err := base64.StdEncoding.DecodeString(encrypted.AuthTag)
	if err != nil {
		return nil, fmt.Errorf("failed to decode auth tag: %w", err)
	}

	// Create AES cipher
	block, err := aes.NewCipher(key)
	if err != nil {
		return nil, fmt.Errorf("failed to create cipher: %w", err)
	}

	// Create GCM
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return nil, fmt.Errorf("failed to create GCM: %w", err)
	}

	// Combine ciphertext and auth tag (GCM expects them concatenated)
	ciphertextWithTag := append(ciphertext, authTag...)

	// Decrypt
	plaintext, err := gcm.Open(nil, iv, ciphertextWithTag, nil)
	if err != nil {
		return nil, fmt.Errorf("decryption failed (authentication failed or corrupted data): %w", err)
	}

	return plaintext, nil
}

// EncryptData encrypts data using AES-256-GCM
//
// Returns an EncryptedPayload containing the ciphertext, IV, and auth tag.
func EncryptData(data interface{}, key []byte) (EncryptedPayload, error) {
	// Serialize data to JSON
	plaintext, err := json.Marshal(data)
	if err != nil {
		return EncryptedPayload{}, fmt.Errorf("failed to marshal data: %w", err)
	}

	// Create AES cipher
	block, err := aes.NewCipher(key)
	if err != nil {
		return EncryptedPayload{}, fmt.Errorf("failed to create cipher: %w", err)
	}

	// Create GCM
	gcm, err := cipher.NewGCM(block)
	if err != nil {
		return EncryptedPayload{}, fmt.Errorf("failed to create GCM: %w", err)
	}

	// Generate random IV
	iv := make([]byte, gcm.NonceSize())
	if _, err := rand.Read(iv); err != nil {
		return EncryptedPayload{}, fmt.Errorf("failed to generate IV: %w", err)
	}

	// Encrypt
	ciphertextWithTag := gcm.Seal(nil, iv, plaintext, nil)

	// GCM appends the auth tag to the ciphertext
	// Split them for separate base64 encoding
	tagSize := gcm.Overhead() // 16 bytes for GCM
	ciphertext := ciphertextWithTag[:len(ciphertextWithTag)-tagSize]
	authTag := ciphertextWithTag[len(ciphertextWithTag)-tagSize:]

	return EncryptedPayload{
		Ciphertext: base64.StdEncoding.EncodeToString(ciphertext),
		IV:         base64.StdEncoding.EncodeToString(iv),
		AuthTag:    base64.StdEncoding.EncodeToString(authTag),
		Algorithm:  "AES-256-GCM",
	}, nil
}

// DecryptRequest decrypts an encrypted API request into the target type
//
// This is a convenience wrapper around DecryptPayload that also handles unmarshaling.
func DecryptRequest[T any](req EncryptedRequest, key []byte) (T, error) {
	var result T

	plaintext, err := DecryptPayload(req.Encrypted, key)
	if err != nil {
		return result, err
	}

	if err := json.Unmarshal(plaintext, &result); err != nil {
		return result, fmt.Errorf("failed to unmarshal decrypted data: %w", err)
	}

	return result, nil
}

// EncryptResponse encrypts a response for the client
//
// This is a convenience wrapper around EncryptData.
func EncryptResponse(data interface{}, key []byte) (EncryptedResponse, error) {
	encrypted, err := EncryptData(data, key)
	if err != nil {
		return EncryptedResponse{}, err
	}

	return EncryptedResponse{
		Encrypted: encrypted,
	}, nil
}

// CleanupMemory forces garbage collection and clears sensitive data
//
// Call this after processing each request to ensure ephemeral processing.
func CleanupMemory() {
	runtime.GC()
}

// GetEncryptionKey retrieves the encryption key from environment variables
//
// In production, this should be set via Vercel environment variables.
// For development, a default key is used (INSECURE - do not use in production).
func GetEncryptionKey() ([]byte, error) {
	// Try to get from environment variable
	keyHex := os.Getenv("ANALYTICS_ENCRYPTION_KEY")

	if keyHex == "" {
		// Development mode: use a fixed key (INSECURE)
		// In production, this should NEVER be used
		if os.Getenv("VERCEL_ENV") == "production" {
			return nil, errors.New("ANALYTICS_ENCRYPTION_KEY must be set in production")
		}

		// 32-byte (256-bit) key for development only
		return []byte("dev-key-32-bytes-do-not-use!"), nil
	}

	// Decode hex key
	key, err := hex.DecodeString(keyHex)
	if err != nil {
		return nil, fmt.Errorf("invalid encryption key format: %w", err)
	}

	// Verify key length (256 bits = 32 bytes)
	if len(key) != 32 {
		return nil, fmt.Errorf("invalid key length: expected 32 bytes, got %d", len(key))
	}

	return key, nil
}
